server/src/main/java/org/apache/druid/segment/realtime/appenderator/BaseAppenderatorDriver.java

```diff
@@ -652,19 +656,16 @@ ListenableFuture<SegmentsAndCommitMetadata> publishInBackground(
                   segmentsAndCommitMetadata.getSegments().forEach(dataSegmentKiller::killQuietly);
                 }
               } else {
+                log.errorSegments(ourSegments, "Failed to publish segments");
+                if (publishResult.getErrorMsg() != null && publishResult.getErrorMsg().contains(("Failed to update the metadata Store. The new start metadata is ahead of last commited end state."))) {
+                  throw new ISE(publishResult.getErrorMsg());
+                }
                 // Our segments aren't active. Publish failed for some reason. Clean them up and then throw an error.
                 segmentsAndCommitMetadata.getSegments().forEach(dataSegmentKiller::killQuietly);
```

Comment:
Segments should be cleaned up if the retries are exhausted or an unrecoverable exception is thrown.

Reply:
For this, I will need to add the RetryUtils.retry() in the try/catch block, right? or is there a another way of doing this

Reply:
Yes